common/definition/deduplicationKeys.go

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package definition
+
+import (
+	"fmt"
+)
+
+const (
+	keySeparator     = "::"
+	keyValue         = "%v"
+	cacheKeyTemplate = keyValue + keySeparator + keyValue
+)
+
+type (
+	DeduplicationKey interface {
+		String() string
+	}
+)
+
+// Deduplication key type
+const (
+	reapplyEventCacheKey = iota
+)
+
+// Deduplication key struct
+type (
+	ReapplyEventKey struct {
+		runID   string
+		eventID int64
+		version int64
+	}
+)
+
+func NewReapplyEventKey(
+	runID string,
+	eventID int64,
+	version int64,
+) ReapplyEventKey {
+
+	return ReapplyEventKey{
+		runID:   runID,
+		eventID: eventID,
+		version: version,
+	}
+}
+
+func (e ReapplyEventKey) String() string {
+	return fmt.Sprintf(keyValue+keySeparator+keyValue+keySeparator+keyValue, e.runID, e.eventID, e.version)
+}
```

Comment:
i guess GetID is preferable

Reply:
and we should generate the ID during initialization instead of calculation each time

Reply:
done